File: 0020-valid-parentheses/0020-valid-parentheses.py
# time complexity: O(n)
# space complexity: O(n)
class Solution:
    def isValid(self, s: str) -> bool:
        bracketMap = {"(": ")", "[": "]", "{": "}"}
        openSet = set(["(", "[", "{"])
        stack = []
        for char in s:
            if char in openSet:
                stack.append(char)
            elif stack and char == bracketMap[stack[-1]]:
                stack.pop()
            else:
                return False
        return stack == []

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c in '{[(':
                stack.append(c)
            elif c == ')' and stack and stack[-1] == '(':
                stack.pop()
            elif c == ']' and stack and stack[-1] == '[':
                stack.pop()
            elif c == '}' and stack and stack[-1] == '{':
                stack.pop()
            else:
                return False
        return len(stack) == 0

s = "()"
print(Solution().isValid(s))
s = "()[]{}"
print(Solution().isValid(s))
s = "(]"
print(Solution().isValid(s))
s = "([])"
print(Solution().isValid(s))
s = "([)]"
print(Solution().isValid(s))

File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 * Time complexity: O(n)
 * Space complexity: O(n)
 */
var isValid = function(s) {
    const bracketMap = { "(": ")", "[": "]", "{": "}" };
    const openSet = new Set(["(", "[", "{"]);
    const stack = [];
    
    for (let char of s) {
        if (openSet.has(char)) {
            stack.push(char);
        } else if (stack.length > 0 && char === bracketMap[stack[stack.length - 1]]) {
            stack.pop();
        } else {
            return false;
        }
    }
    
    return stack.length === 0;
};

const input = "(())";
console.log(isValid(input)); // Output will be true or false based on the validity of the parentheses.
